datafusion/core/src/datasource/memory.rs

```diff
@@ -333,7 +337,11 @@ impl DisplayAs for MemSink {
 }
 
 impl MemSink {
+    /// Creates a new [`MemSink`].
+    ///
+    /// The caller is responsible for ensuring that there is at least one partition to insert into.
```

Comment:
I think you could avoid the assert / panic here by changing the `new` method to `try_new` and return a `Result` so it could be an error if the batches is empty (and move the check to try_new)

However, I agree this PR is an improvement over what is currently on main so this PR is good to merge as is.

ALso I double checked and `MemSink` is not part of the Public API so this seems reasonable to me

Reply:
In hindsight this would be the better API, even though its just internal.
As I don't have access to a computer for now, I'd merge it like that and if the API ever becomes public we expose the error to the caller.

Reply:
Merged. Thanks @tobixdev . I will create a new PR to follow @alamb 's suggestion.